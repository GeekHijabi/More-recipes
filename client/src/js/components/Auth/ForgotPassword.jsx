import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import RecipeHeader from '../Partials/RecipeHeader';
import { apiForgotPassword } from '../../actions/auth';
import
validateforgotPasswordInput from
  '../../utils/validations/forgotPasswordValidation';

/**
 * @class ForgotPassword
 * @extends {React.Component}
 */
export class ForgotPassword extends React.Component {
  /**
   * @description Constructor function
   * @param {any} props
   * @memberof SignIn
   * @return {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      errors: {}
    };
    this.onChange = this.onChange.bind(this);
    this.onClick = this.onClick.bind(this);
  }

  /**
   *
   * @param {any} event
   * @memberof ForgotPassword
   * @returns {void}
 */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  /**
   *
   * @param {any} event
   * @memberof ForgotPassword
   * @returns {void}
 */
  onClick(event) {
    if (this.isValid()) {
      event.preventDefault();
      this.props.apiForgotPassword(this.state.email)
        .then((res) => {
          toastr.options = {
            closeButton: true,
            progressBar: true
          };
          toastr.success(res.data.message);
          this.props.history.push('/signin');
        })
        .catch((err) => {
          toastr.danger(err.data);
        });
      this.setState({ email: '' });
    }
  }

  /**
  *
  * @param {any} event
  * @memberof SignUp
  * @returns {object} event
  */
  isValid() {
    const { errors, isValid } = validateforgotPasswordInput(this.state);

    if (!isValid) {
      this.setState({ errors });
    }

    return isValid;
  }

  /**
   * @description Constructor function
   * @param {any} props
   * @memberof ForgotPassword
   * @return {void}
   */
  render() {
    const { errors } = this.state;
    return (
      <div page="signin">
        <div className="overlay" />
        <RecipeHeader />
        <div className="card card-style reset-card">
          <div className="card-body mx-4">
            <div className="text-center">
              <h3 className="dark-grey-text mb-4" >
                <strong>Forgot password?</strong>
              </h3>
              <span className="mb-3 reset-text">No worries!
              Input your email address below and we will send you
              reset information
              </span>
            </div>
            <div className="md-form">
              <label htmlFor="Form-email1" className="form-label">
                <input
                  type="text"
                  id="Form-email1"
                  className="form-control"
                  placeholder="Email"
                  name="email"
                  onChange={this.onChange}
                  value={this.state.email}
                  required
                />
                {errors.email &&
                <small style={{ color: '#A43741' }}>
                  {errors.email }
                </small>}
              </label>
            </div>
            <div className="text-center mb-3">
              <button
                type="button"
                className="btn btn-white btn-rounded"
                onClick={this.onClick}
              >
                       Send me an instruction
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
ForgotPassword.propTypes = {
  apiForgotPassword: PropTypes.func.isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired
};
export default connect(null, { apiForgotPassword })(ForgotPassword);
