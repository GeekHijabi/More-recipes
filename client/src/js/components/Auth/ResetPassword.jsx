import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import RecipeHeader from '../Partials/RecipeHeader';
import { apiResetPassword } from '../../actions/auth';
import
validateresetPasswordInput from
  '../../utils/validations/resetPasswordValidation';

/**
 * @class ResetPassword
 * @extends {React.Component}
 */
export class ResetPassword extends React.Component {
  /**
   * @description Constructor function
   * @param {any} props
   * @memberof SignIn
   * @return {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      newPassword: '',
      confirmPassword: '',
      errors: {}
    };
    this.onChange = this.onChange.bind(this);
    this.onClick = this.onClick.bind(this);
  }

  /**
   *
   * @param {any} event
   * @memberof ResetPassword
   * @returns {void}
 */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  /**
   *
   * @param {any} event
   * @memberof ResetPassword
   * @returns {void}
 */
  onClick(event) {
    const baseUrl = window.location.href;
    const url = baseUrl.split('?token=');
    const currentURL = url[0];
    const userId = currentURL.split('/')[currentURL.split('/').length - 1];
    if (this.isValid()) {
      event.preventDefault();
      this.props.apiResetPassword(userId, this.state.newPassword, url[1])
        .then((data) => {
          toastr.options = {
            closeButton: true,
            progressBar: true
          };
          toastr.success(data.newPasswordMessage);
          this.props.history.push('/signin');
        })
        .catch((err) => {
          toastr.danger(err.data);
        });
    }
  }

  /**
  *
  * @param {any} event
  * @memberof ResetPassword
  * @returns {object} event
  */
  isValid() {
    const { errors, isValid } = validateresetPasswordInput(this.state);

    if (!isValid) {
      this.setState({ errors });
    }

    return isValid;
  }

  /**
   * @description Constructor function
   * @param {any} props
   * @memberof ResetPassword
   * @return {void}
   */
  render() {
    const { errors } = this.state;
    return (
      <div page="signin">
        <div className="overlay" />
        <RecipeHeader />
        <div className="card card-style">
          <div className="card-body mx-4">
            <div className="text-center">
              <h3 className="dark-grey-text mb-5" >
                <strong>Reset password</strong>
              </h3>
            </div>
            <div className="md-form">
              <label htmlFor="Form-email1" className="form-label">
                <input
                  type="password"
                  className="Form-email1 form-control"
                  placeholder="New Password"
                  name="newPassword"
                  onChange={this.onChange}
                  value={this.state.newPassword}
                  required
                />
                {errors.password &&
                <small style={{ color: '#A43741' }}>
                  {errors.password }
                </small>}
              </label>

              <label htmlFor="Form-email1" className="form-label">
                <input
                  type="password"
                  className="Form-email1 form-control"
                  placeholder="confirm Password"
                  name="confirmPassword"
                  onChange={this.onChange}
                  value={this.state.confirmPassword}
                  required
                />
                {errors.confirmPassword &&
                <small style={{ color: '#A43741' }}>
                  {errors.confirmPassword }
                </small>}
              </label>
            </div>
            <div className="text-center mb-3">
              <button
                type="button"
                className="btn btn-white btn-rounded"
                onClick={this.onClick}
              >
                        Reset password
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
ResetPassword.propTypes = {
  apiResetPassword: PropTypes.func.isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired
};
export default connect(null, { apiResetPassword })(ResetPassword);
