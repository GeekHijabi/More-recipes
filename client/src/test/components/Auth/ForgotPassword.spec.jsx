import React from 'react';
import { shallow } from 'enzyme';
import { ForgotPassword } from '../../../js/components/Auth/ForgotPassword';

const userInput = {
  email: 'test@tester.com'
};

const props = {
  apiForgotPassword: jest.fn(() => Promise.resolve({ ok: true })),
  history: {
    push: () => Promise.resolve({ ok: true }),
  }
};
describe('Component: ForgotPassword', () => {
  beforeEach(() => {
    global.toastr = {
      success: () => {},
      error: () => {}
    };
  });

  describe('forgot password in snapshot', () => {
    it('should render correctly', () => {
      const forgotPassword = shallow(<ForgotPassword {...props} />);
      expect(forgotPassword).toMatchSnapshot();
    });
  });

  describe('onChange', () => {
    it('should change event', () => {
      const event = {
        target: { name: 'email', value: 'tester@test.com' },
      };
      const forgotPassword = shallow(<ForgotPassword {...props} />);
      forgotPassword.instance().onChange(event);
      expect(forgotPassword.instance().state.email).toEqual('tester@test.com');
    });
  });

  describe('onClick', () => {
    it('should login user when user details is set to the state', () => {
      const event = {
        preventDefault: jest.fn()
      };
      const submit = shallow(<ForgotPassword {...props} />);
      submit.setState(userInput);
      submit.setState({ hasError: true, errorMessage: 'error', errors: {} });
      submit.instance().onClick(event);
    });
  });

  describe('isValid', () => {
    it('should check if data is valid', () => {
      const errors = {
        email: ''
      };
      const event = {
        preventDefault: jest.fn()
      };
      const isNotvalid = shallow(<ForgotPassword {...props} />);
      // isNotvalid.setState(userInput);
      isNotvalid.setState(errors);
      isNotvalid.instance().isValid(event);
    });
  });
});

