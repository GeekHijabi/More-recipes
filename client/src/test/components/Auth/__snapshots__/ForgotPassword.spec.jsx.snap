// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ForgotPassword forgot password in snapshot should render correctly 1`] = `
<div
  page="signin"
>
  <div
    className="overlay"
  />
  <withRouter(Connect(RecipeHeader)) />
  <div
    className="card card-style"
  >
    <div
      className="card-body mx-4"
    >
      <div
        className="text-center"
      >
        <h3
          className="dark-grey-text mb-5"
        >
          <strong>
            Forgot password?
          </strong>
        </h3>
      </div>
      <div
        className="md-form"
      >
        <label
          className="form-label"
          htmlFor="Form-email1"
        >
          Email
          <input
            className="form-control"
            id="Form-email1"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            required={true}
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        className="text-center mb-3"
      >
        <button
          className="btn btn-white btn-rounded"
          onClick={[Function]}
          type="button"
        >
          Send me an instruction
        </button>
      </div>
    </div>
  </div>
</div>
`;
